Add a default optimization level to generate_preset_pass_manager

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a default value to the generate_preset_pass_manager's optimization_level argument. If it's not specified optimization level 2 will be used. After #12148 optimization level 2 is a better fit for an optimal tradeoff between heuristic effort and runtime that makes it well suited as a default optimization level.

### Details and comments

mtreinish committed:
Add a default optimization level to generate_preset_pass_manager

This commit adds a default value to the generate_preset_pass_manager's
optimization_level argument. If it's not specified optimization level 2
will be used. After #12148 optimization level 2 is a better fit for an
optimal tradeoff between heuristic effort and runtime that makes it
well suited as a default optimization level.

kevinhartman commented on qiskit/transpiler/preset_passmanagers/__init__.py:
```diff
@@ -205,6 +207,16 @@ def generate_preset_pass_manager(
+    # Handle positional arguments for target and backend. This enables the usage
+    # pattern `generate_preset_pass_manager(backend.target)` to generate a default
+    # pass manager for a given target.
+    if isinstance(optimization_level, Target):
```
Personally, I'd rather this function not support any positional arguments. 

I think we want a more ergonomic API on top of this function eventually anyways, e.g. https://github.com/Qiskit/qiskit/pull/12161, so it seems too kludgy to me to try and make this accept a target positionally here.

mtreinish replied:
#12161 is actually what I'm not sure we want to do. Adding a 3rd API with different semantics to do what we already have two interfaces for seems like a mistake to me. If people really want a different name I feel like we really should just alias it (but I still don't think it's worth it). But adding yet another entrypoint to accomplish the same thing feels like a mistake. If people are complaining about the ergonomics of the existing interface I feel like we should just evolve it in-place instead of diverging it again and requiring people to learn yet another thing.

jakelishman replied:
Yeah, it's unfortunate that we've ended up here, but I think the right thing to do for users is to make it so the common patterns `{transpile,generate_preset_pass_manager}({backend,target}, optimization_level=2)` work the same in both forms.  It ends up in an ugly signature for us, but we can tidy that up in place and potentially fix the signature properly for Qiskit 2.0+.

jakelishman reviewed (commented):
This looks fine in principle to me, I'm just concerned about how we're setting this to have a different default optimisation level to `transpile`.  We already get complaints whenever the interfaces of `transpile` and this diverge around type inference / whatever, and I guess I'd roughly rather we moved the two in step?

I can live with the split if it was already discussed more in the team, though.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into default-generate-preset-pass-manager

mtreinish committed:
Update transpile()'s default opt level to match

This commit updates the transpile() function's optimization_level argument
default value to match generate_preset_pass_manager's new default to use 2
instead of 1. This is arguably a breaking API change, but since the
semantics are equivalent with two minor edge cases with implicit behavior
that were a side effect of the level 1 preset pass manager's construction
(which are documented in the release notes) we're ok making it in this
case. Some tests which we're relying on the implicit behavior of
optimization level 1 are updated to explicitly set the optimization
level argument which will retain this behavior.

mtreinish commented:
> This looks fine in principle to me, I'm just concerned about how we're setting this to have a different default optimisation level to `transpile`. We already get complaints whenever the interfaces of `transpile` and this diverge around type inference / whatever, and I guess I'd roughly rather we moved the two in step?
> 
> I can live with the split if it was already discussed more in the team, though.

I updated the `transpile()` default to level 2 to match in: https://github.com/Qiskit/qiskit/pull/12150/commits/34e9ee4977307383764df3e5545975ec003705ca I'm a bit worried this is a breaking api change though. While at a high level the semantics are equivalent as the transpile function compiles the circuit in the same way. There were some implicit behaviors that came with level 1 which don't work in higher optimization levels. The two cases I know about are the implicit trivial layout if it's perfect and working with discrete basis sets. The unit tests flagged these as they were relying on this behavior, I updated those tests to explicitly use level 1 in those cases. I've also documented these edge cases in the release notes. It's arguable whether that's is part of our API contract though, because neither are actually defined behavior, more just implicit expectations/assumptions. However, level 1 being the default was explicitly documented...

mtreinish committed:
Update more tests expecting optimization level 1

mtreinish commented:
I'm thinking that if we want to do this for 1.1.0 we should drop https://github.com/Qiskit/qiskit/commit/34e9ee4977307383764df3e5545975ec003705ca and save that for 1.2. There seem to be a lot of issues with using level2 as the default in testing. There might be some more implicit assumptions I didn't think of that the tests are flagging. FWIW, every commit I've pushed up I ran the tests locally with and everything passed I assume it's either the presence of optional dependencies or py3.8 vs 3.12 that's causing the failures. But it's making me uneasy about making a potentially breaking change at the last minute before a release.

jakelishman commented:
An alternative to moving `transpile`'s default in this PR is to set the default for `generate_preset_pass_manager` to O1, and then move them both to O2 for Qiskit 2.0?  Totally agree that it feels too big a step to move `transpile`'s default at this point in time.

I'd like to see them stay in step at all times, but I can live with it if the decision is to go with O2 for this and move `transpile` to match it later.

ElePT committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit into default-generate-preset-pass-manager

ElePT committed:
* Set optimization level to 1 in test_approximation_degree.

* Replace use of transpile with specific pass in  HLS tests.

ElePT committed:
Set optimization_level=1 in layout-dependent tests.

ElePT reviewed (commented):
The PR looks good to me, I went in and fixed a couple of missing tests that were optimization-level dependant, or were using the full pipeline to only test the effect of a single transpiler pass (and that made them optimization-level dependant). Before approving, I want to confirm that we want to move on with the optimization level 2 default in both `transpile` and `generate_ppm` (I am going to call it like this from now on) for 1.2. Thoughts? @jakelishman @mtreinish @1ucian0

ElePT committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit into default-generate-preset-pass-manager

mtreinish commented:
I'm good with doing this for 1.2, if there are no objections from anyone else. I'm personally still a bit apprehensive about the API implications on changing `transpile()` because of the test changes we needed to make and I'm concerned those are implicit user expectations even if they're not explicitly documented around the behavior of `transpile(qc, backend)`. But I'm willing to fall back on the statement that our stability contract is to our documented interfaces, and these underlying assumptions aren't documented as part of `transpile()`. Then this can be viewed as being no different than changing an optimization pass to be do more work in the preset pass managers.

I will push up a small tweak to the upgrade release note though to elaborate a bit more on the benefits of the change to `transpile()` since I just realized it's never mentioned.

mtreinish committed:
Expand upgrade note explanation on benefits of level 2

ElePT committed:
Merge branch 'main' into default-generate-preset-pass-manager

ElePT commented on releasenotes/notes/default-level-2-generate-preset-passmanager-ec758ddc896ae2d6.yaml:
```diff
@@ -0,0 +1,44 @@
+    level 2 provides a much better balance between the run time of the function and the optimizations it
+    performs, it's a better tradeoff to use by default.
+
+    The API of :func:`.transpile` remains unchange because fundamentally level 2 and level 1
```
```suggestion
    The API of :func:`.transpile` remains unchanged because, fundamentally, level 2 and level 1
```

ElePT commented on releasenotes/notes/default-level-2-generate-preset-passmanager-ec758ddc896ae2d6.yaml:
```diff
@@ -0,0 +1,44 @@
+    The only potential issue is that if you were relying on an implicit trivial layout (where qubit 0
+    in the circuit passed to :func:`.transpile` is mapped to qubit 0 on the target backend/coupling,
+    1->1, 2->2, etc.) without specifying ``optimization_level=1``, ``layout_method="trivial"``, or
+    explicitly setting ``initial_layout`` when calling :func:`.transpile`. This behavior was a side
```
```suggestion
    The only potential issue is that your transpilation workflow may be relying on an implicit trivial layout (where qubit 0
    in the circuit passed to :func:`.transpile` is mapped to qubit 0 on the target backend/coupling,
    1->1, 2->2, etc.) without specifying ``optimization_level=1``, ``layout_method="trivial"``, or
    explicitly setting ``initial_layout`` when calling :func:`.transpile`. This behavior was a side
```

ElePT commented on releasenotes/notes/default-level-2-generate-preset-passmanager-ec758ddc896ae2d6.yaml:
```diff
@@ -0,0 +1,44 @@
+    1->1, 2->2, etc.) without specifying ``optimization_level=1``, ``layout_method="trivial"``, or
+    explicitly setting ``initial_layout`` when calling :func:`.transpile`. This behavior was a side
+    effect of the preset pass manager construction in optimization level 1 and is not mirrored in
+    level 2. If you need this behavior you can use any of the three options list previously to make
```
```suggestion
    level 2. If you need this behavior you can use any of the three options listed previously to make
```

ElePT commented on releasenotes/notes/default-level-2-generate-preset-passmanager-ec758ddc896ae2d6.yaml:
```diff
@@ -0,0 +1,44 @@
+    this behavior explicit.
+
+    Similarly if you were targeting a discrete basis gate set you may encounter an issue using the
+    new default with optimization level 2 (or 3) as the additional optimization passes that run in
```
```suggestion
    Similarly, if you were targeting a discrete basis gate set you may encounter an issue using the
    new default with optimization level 2 (or running explicitly optimization level 3), as the additional optimization passes that run in
```

ElePT reviewed (commented):
I left a few suggestions for the reno.

ElePT committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit into default-generate-preset-pass-manager

ElePT committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit into default-generate-preset-pass-manager

ElePT committed:
Merge branch 'default-generate-preset-pass-manager' of https://github.com/mtreinish/qiskit-core into default-generate-preset-pass-manager

ElePT committed:
Apply Elena's reno suggestions

ElePT reviewed (approved):
I commited the reno suggestions and will approve the PR, @mtreinish, feel free to add it to the merge queue if you agree with the changes or modify them if you don't.
